### PR TITLE
fix: Table Widget Icon Button should be having a default variant

### LINKED ISSUE #8527: [Bug]: Table Widget Icon Button should be having a default variant
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Icon Button in the table widget should be having a default button variant selection.
It should be selected as PRIMARY by default and the Icon button should look like a PRIMARY button.
Right now, the variant selection is empty and the icon button is behaving like a TERTIARY button.
This is going to create inconsistencies in the future.

Screenshot 2021-10-14 at 12 15 23 PM

### Steps To Reproduce

```markdown
1. Add a table widget
2. Change column type to "Icon Button"
3. Select an icon
4. See the Button Variant selection and the icon button in the table
```

### Environment

Production

### Version

Cloud

### PR DESCRIPTION
**Pull request by wmdev0808**
## Description
Icon Button in the table widget should be having a default button variant selection.
It should be selected as PRIMARY by default and the Icon button should look like a PRIMARY button.
Right now, the variant selection is empty and the icon button is behaving like a TERTIARY button.
This is going to create inconsistencies in the future.

Fixes #8527 

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

- Test A
- Test B

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: fix/8527-table-icon-button-default-variant 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 54.64 **(0.01)** | 36.37 **(0.02)** | 33.1 **(0.01)** | 55.16 **(0.01)**
 :green_circle: | app/client/src/utils/DSLMigrations.ts | 73.74 **(0.12)** | 65.09 **(-0.05)** | 70.21 **(0)** | 73.55 **(0.12)**
 :green_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.26 **(0.22)** | 41.25 **(0.83)** | 34.48 **(0)** | 56.22 **(0.26)**
 :red_circle: | app/client/src/utils/migrations/TableWidget.ts | 84.03 **(-0.79)** | 75 **(-0.76)** | 90 **(-2.59)** | 85.65 **(-1.08)**
 :red_circle: | app/client/src/widgets/TableWidget/widget/index.tsx | 26.67 **(0)** | 6.02 **(-0.03)** | 28 **(0)** | 26.56 **(0)**</details>

### CODE REVIEW

**Review by aswathkk — CHANGES_REQUESTED** [review]
@wmdev0808 I've discussed with @riodeuno and found that the ideal way to provide default value for buttonVariant would be to provide it inside [`getTableColumns()`](https://github.com/appsmithorg/appsmith/blob/77e3c4f4db24fa65a05aa834902d9bd7a66e89a0/app/client/src/widgets/TableWidget/widget/index.tsx#L243) function.

**Review by aswathkk — CHANGES_REQUESTED** [review]
I think we need a migration here

Release:
Screenshot 2021-11-19 at 12 26 45 PM

After this PR:
Screenshot 2021-11-19 at 12 26 55 PM

File: app/client/src/utils/migrations/TableWidget.ts [hunk 1]
```diff
@@ -462,3 +462,23 @@ export const migrateTableSanitizeColumnKeys = (currentDSL: DSLWidget) => {
 
   return currentDSL;
 };
+
+export const migrateTableWidgetIconButtonVariant = (currentDSL: DSLWidget) => {
+  currentDSL.children = currentDSL.children?.map((child: WidgetProps) => {
+    if (child.type === "TABLE_WIDGET") {
+      child.primaryColumns.forEach((accessor: string) => {
```
Comment (aswathkk) [inline app/client/src/utils/migrations/TableWidget.ts:469]: `child.primaryColumns` is an object and it doesn't have `forEach` method.

Comment (aswathkk) [reply app/client/src/utils/migrations/TableWidget.ts:469]: This error is being thrown at Deploy preview

**Review by aswathkk — CHANGES_REQUESTED** [review]
@wmdev0808 Have you tried to see the Deploy preview?
It is throwing errors on migration.

**Review by aswathkk — CHANGES_REQUESTED** [review]
@wmdev0808 If you do not alter buttonVariant of an icon button on release, you will see it will be styled as "TERTIARY" button. We want to keep that behaviour for existing users. That's why migration is required.

For more info, check out the [previous comment](https://github.com/appsmithorg/appsmith/pull/8664#pullrequestreview-810814808) to see how my application's icon button behaved when I switched from release to this PR.

File: app/client/src/utils/migrations/TableWidget.ts [hunk 2]
```diff
@@ -462,3 +462,27 @@ export const migrateTableSanitizeColumnKeys = (currentDSL: DSLWidget) => {
...
+      const primaryColumns = child.primaryColumns as Record<
+        string,
+        ColumnProperties
+      >;
+      Object.keys(primaryColumns).forEach((accessor: string) => {
+        const primaryColumn = primaryColumns[accessor];
+
+        if (primaryColumn.columnType === "iconButton") {
+          if (!("buttonVariant" in primaryColumn)) {
+            primaryColumn.buttonVariant = "PRIMARY";
```
[the remark this answers is not in the document]
Comment (wmdev0808) [reply app/client/src/utils/migrations/TableWidget.ts:478]: I can't understand your point.
My code is just applying changes if the current table column has no buttonVariant property, so it should be PRIMARY by default

**Review by aswathkk — APPROVED** [review]

**Pull request merged**